### ISSUE TITLE
Data Deletion from PostHog's databases

### ISSUE THREAD

**Issue by tiina303** [PostHog/posthog#20221]
## Is your feature request related to a problem?

PII (Personal Identifiable Information) can be sent to PostHog and end up being stored in our databases. Users need a way to be able to request deletion of that data.
 
The api and UI allow triggering data deletions on a person level, however the deletion system is currently experiencing difficulties & might not cover data in all the systems, e.g. in addition to events data the distinct_id itself could also contain PII.

Secondly sometimes users would like to delete PII based on a time range, e.g. there was an implementation error that for a couple of minutes sent PII, then we'd ideally delete data only during that time frame and not the full history.

## Describe the solution you'd like

1. Fix the person level data deletion system & improve its reliability/monitoring/alerting.
    1. Ensure related data gets deleted from all systems
2. Provide a solution for time based data deletion

## Describe alternatives you've considered

PostHog has events deduplication, so users could consider sending updated events with the same uuid without PII. The problem with this approach is that
1. ClickHouse deduplication is eventual, so there isn't a guarantee for when the data would get deleted
2. Re-submitting events could have unintended side-effects, e.g. user properties are set at the time of ingestion, so sending a day old event again might override the person properties that had been updated afterwards
3. it would be quite cumbersome to re-submit a lot of events and that would cause increased load on our ingestion system

## Additional context

Data modification is a very complex operation, so we don't plan to offer that at this point in time.

Internal notes: https://github.com/orgs/PostHog/projects/98/views/1?pane=issue&itemId=51927319

**Comment by tkaemming** [comment]
Also, I'm not sure if it make sense to be part of this ticket or better kept as something separate (but related): [event deletions for persons right now delete based on the `person_id`](https://github.com/PostHog/posthog/blob/08acfba72ee852c6ac5af957223447ccf6457de0/posthog/models/async_deletion/delete_events.py#L96) which is going to potentially be out-of-date if the person hasn't been fully squashed across all partitions, allowing events that should have been deleted to stick around.

**Comment by bretthoerner** [comment]
> Also, I'm not sure if it make sense to be part of this ticket or better kept as something separate (but related): [event deletions for persons right now delete based on the `person_id`](https://github.com/PostHog/posthog/blob/08acfba72ee852c6ac5af957223447ccf6457de0/posthog/models/async_deletion/delete_events.py#L96) which is going to potentially be out-of-date if the person hasn't been fully squashed across all partitions, allowing events that should have been deleted to stick around.

Good call out, maybe the new JOIN shenanigans could use all of the Person's distinct_ids?

**Comment by tkaemming** [comment]
> Good call out, maybe the new JOIN shenanigans could use all of the Person's distinct_ids?

Yeah, seems like it'd be possible to use the overrides table here as well (at least in theory, I'm not familiar enough offhand with the order-of-operations during deletions here to know if that's totally accurate as-is.)

**Comment by mateozaratefw** [comment]
Any updates on this? 
I need to delete a ton of user sessions and the UI is awful. The API doesn't seems to have any endpoint as well!!
